connmgr/connmgr.go

```diff
@@ -14,14 +14,23 @@ import (
 	"github.com/libp2p/go-libp2p-core/peer"
 )
 
-// ConnManager tracks connections to peers, and allows consumers to associate metadata
-// with each peer.
+// SupportsDecay evaluates if the provided ConnManager supports decay, and if
+// so, it returns the Decayer object. Refer to godocs on Decayer for more info.
+func SupportsDecay(mgr ConnManager) (Decayer, bool) {
```

Comment:
Shouldn't we simply have the extra methods and cast/check the interface?
It feels more idiomatic.

Reply:
Isn't that what this does? In the implementation we use type embedding, that's why this works.

Reply:
Fair enough, it amounts to the same thing.
Perhaps we could name the method better though, how about just `Decayer() -> (Decayer, bool)`?

Reply:
Name collision, though. In other languages you’d do: `connMgr.as(Decayer.class)`, but no generics :-( I considered CastAsDecayer, ToDecayer, before settling for SupportsDecayer.

Reply:
duh.... How about simply `GetDecayer`?

Reply:
I don’t think it makes a substantial difference. “Supports” is more correct because the underlying conn mgr may not support decaying.

Reply:
ok, fine by me.